src/seaofnodes/seaOfNodes__Type.ml

```diff
@@ -0,0 +1,208 @@
+(** The module [Identified] is here to hide the tags in the SON types *)
+(** We use a Weak set to remember previously created values,
+    in which we store elements of type Identified t *)
+(** Identified types are just an element with an integer *)
+module type IDENTIFIED = sig
+  type elt
+  type t = int * elt
+  val hash : t -> int
+  val equal : t -> t -> bool
+  val get_value : t -> elt
+  val get_id : t -> int
+end
+module Identified (M : sig
+    type t
+    val hash : t -> int
+    val equal : t -> t -> bool
+  end) : IDENTIFIED with type elt = M.t = struct
+  type t = int * M.t
+  type elt = M.t
+  let hash (_, x) = M.hash x
+  let equal (_, x) (_, y) = M.equal x y
+
+  let get_value (_, x) = x
+  let get_id (id, _) = id
+end
+
+
+module Binop : sig
+  type t = Add
+  val hash : t -> int
+end = struct
+  type t = Add
+  let hash _ = 1
+end
+
+
+module rec JumpSet : GenericSet.GenericSetSig = GenericSet.Make(Jump)
+and BranchSet : GenericSet.GenericSetSig = GenericSet.Make(Branch)
+
+and Data : sig
+  (** It is possible to pattern match elements of type t, *)
+  (** but you must use the smart constructors to build values *)
+  type t = private Const of int | BinOp of {op: Binop.t; operand1: t; operand2: t} | Phi of Phi.t
+
+  val hash : t -> int
+  val equal : t -> t -> bool
+
+  val pp_dot : out_channel -> t -> unit
+
+  (** Constructors *)
+  val const : int -> t
+  val binop : Binop.t -> t -> t -> t
+  val phi : Phi.t -> t
+
+  (** Unique id of the node *)
+  val get_id : t -> int
+end = struct
+  type t = Const of int | BinOp of {op: Binop.t; operand1: t; operand2: t} | Phi of Phi.t
+
+  let hash = function
+    | Const n -> n
+    | BinOp {op; operand1; operand2} ->
+      Binop.hash op lxor Data.hash operand1 lxor Data.hash operand2
+    | Phi phi -> Phi.hash phi
```

Comment:
This is not the hash function we want. We don't want a recursive function here. Have a closer look at Filliatre's book (Hashcons chapter).